packages/apollo-server-core/src/runHttpQuery.ts

```diff
@@ -97,19 +101,18 @@ export async function runHttpQuery(handlerArguments: Array<any>, request: HttpQu
         }
       }
 
-      // Shallow clone context for queries in batches. This allows
-      // users to distinguish multiple queries in the batch and to
-      // modify the context object without interfering with each other.
       let context = optionsObject.context;
-      if (isBatch) {
-        context = Object.assign({},  context || {});
+      if (isFunction(context)) {
+        context = context();
+      } else if (isBatch) {
+        context = Object.assign({}, context || {});
```

Comment:
Would it make sense to make this a shallow clone that respects the prototype? Something like `Object.assign(Object.create(Object.getPrototypeOf(context)), context)`? That would solve the issues people are having with class-based contexts. But maybe that is too much magic and we should steer people towards using a function.

Reply:
If we want the default behaviour to be cloning context for batch requests, then your solution will work for more users out of the box!

If anything I'd consider remove cloning because it's not clear it occurs when reading the API (unless i missed something)